File: dungeon_game.py
class Solution:
    def calculateMinimumHP_1(self,dungeon):
        m=len(dungeon)
        n=len(dungeon[0])
        hp=0
        dp=[[0 for i in range(n+1)] for j in range(m+1)]
        dp[m-1][n]=1
        dp[m][n-1]=1
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                dp[i][j]=max(1,mindp[i+1][j]-dungeon[i][j],dp[i][j+1]-dungeon[i][j])
        return dp[0][0]
    def calculateMinimumHP(self,dungeon):
        m=len(dungeon)
        n=len(dungeon[0])
        dp=[[2**20 for i in range(n+1)] for j in range(m+1)]
        dp[m-1][n]=1
        dp[m][n-1]=1
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                need=min(dp[i+1][j],dp[i][j+1])-dungeon[i][j]
                dp[i][j]=1 if need<=0 else need
        return dp[0][0]
        
if __name__=='__main__':
    sol=Solution()
    dungeon=[[5,23,-48,-21,-72,-62,-19,-55,-3,-93,-84,14,-34,46,-82,-46,39,26,47,-71,-46,-3,-59,-93,-13,-72,19,-43,-51,-2,-6,-53,12,-40,15,-94,37,-3,-32,-97]]
    print sol.calculateMinimumHP_1(dungeon)

File: Algorithm/CPP/dungeon_game.cpp
# include <vector>
# include <iostream>
using namespace std;

class Solution {
public:
	int calculateMinimumHP(vector<vector<int>>& dungeon) {
		int m = dungeon.size();
		int n = dungeon[0].size();
		vector<vector<int>> dp(m+1,vector<int>(n+1,INT_MAX));
		dp[m][n - 1] = 1;
		dp[m - 1][n] = 1;
		int need;
		for (int i = m - 1; i >= 0; i--)
			for (int j = n - 1; j >= 0; j--)
			{
				dp[i][j] = max(min(dp[i + 1][j], dp[i][j + 1]) - dungeon[i][j],1);
			}
		return dp[0][0];

	}
};
int main()
{
	Solution sol;
	vector<int> nums = { -2};
	nums = { 3, 2, -2, 4 };
	//nums = { -4, -3 };
	nums = {1,2,3};
	vector < vector<int>> dungeon = {
		{-2,-3,3},
		{-5,-10,1},
		{10,30,-5}
	};
	dungeon = { { -1, 1 } };
	//dungeon = { { -200 } };
	cout << sol.calculateMinimumHP(dungeon);
	
	cout << endl;
	//for (auto k :nums )cout << k << " ";
	system("pause");
	return 0;
}
